File: servers/src/org/xtreemfs/utils/xtfs_stat.java
/*  Copyright (c) 2008 Konrad-Zuse-Zentrum fuer Informationstechnik Berlin.

 This file is part of XtreemFS. XtreemFS is part of XtreemOS, a Linux-based
 Grid Operating System, see <http://www.xtreemos.eu> for more details.
 The XtreemOS project has been developed with the financial support of the
 European Commission's IST program under contract #FP6-033576.

 XtreemFS is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 2 of the License, or (at your option)
 any later version.

 XtreemFS is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with XtreemFS. If not, see <http://www.gnu.org/licenses/>.
 */
/*
 * AUTHORS: Björn Kolbeck (ZIB), Jan Stender (ZIB)
 */

package org.xtreemfs.utils;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Map;

import org.xtreemfs.common.util.OutputUtils;
import org.xtreemfs.foundation.json.JSONParser;
import org.xtreemfs.foundation.json.JSONString;

/**
 *
 * @author bjko
 */
public class xtfs_stat {

    private static final int FILE    = 1;

    private static final int DIR     = 2;

    private static final int SYMLINK = 3;

    /**
     * @param args
     *            the command line arguments
     */
    public static void main(String[] args) {
        try {

            if (args.length < 1) {
                System.out.println("usage: " + xtfs_stat.class.getSimpleName() + " <file>\n");
                System.exit(1);
            }

            String fileName = "";
            for (String arg : args)
                fileName += arg + " ";

            // remove leading and trailing spaces
            fileName = fileName.trim();

            // remove leading and trailing quotes
            if (fileName.charAt(0) == '"' && fileName.charAt(fileName.length() - 1) == '"')
                fileName = fileName.substring(1, fileName.length() - 1);

            // replace all backslashes with spaces
            fileName = fileName.replaceAll("\\\\ ", " ");

            final String format = "%-25s  %s\n";

            // fetch all XtreemFS-related extended attributes
            Map<String, String> attrs = utils.getxattrs(fileName);
            if (attrs == null) {
                System.err.println("file not found: " + fileName);
                return;
            }

            String url = attrs.get("xtreemfs.url");
            if (url == null) {
                System.out.println("'" + fileName
                    + "' is probably not part of an XtreemFS volume (no MRC URL found).");
                return;
            }

            // first, render all general XtreemFS attributes
            System.out.format(format, "filename", new File(fileName).getName());
            System.out.format(format, "XtreemFS URI", url);
            System.out.format(format, "XtreemFS fileID", attrs.get("xtreemfs.file_id"));

            int ftype = Integer.parseInt(attrs.get("xtreemfs.object_type"));
            switch (ftype) {
            case FILE:
                System.out.format(format, "object type", "regular file");
                break;
            case DIR:
                System.out.format(format, "object type", "directory"
                    + (attrs.get("xtreemfs.file_id").endsWith(":1") ? " (volume root)" : ""));
                break;
            case SYMLINK:
                System.out.format(format, "object type", "symlink");
                break;
            }
            System.out.format(format, "owner", attrs.get("xtreemfs.owner"));
            System.out.format(format, "group", attrs.get("xtreemfs.group"));

            if (ftype == FILE) {
                String readOnly = attrs.containsKey("xtreemfs.read_only") ? attrs
                        .get("xtreemfs.read_only") : String.valueOf(false);
                System.out.format(format, "read-only", readOnly);
            }

            // if the file refers to a directory, render directory attributes
            if (ftype == DIR) {
                String defSP = attrs.get("xtreemfs.default_sp");
                if (defSP == null)
                    defSP = "none";
                System.out.format(format, "default striping policy", defSP);
            }

            // render other known XtreemFS attributes
            if (attrs.containsKey("xtreemfs.ac_policy_id"))
                System.out.format(format, "access control policy ID", attrs
                        .get("xtreemfs.ac_policy_id"));

            if (attrs.containsKey("xtreemfs.osdsel_policy_id"))
                System.out.format(format, "OSD selection policy ID", attrs
                        .get("xtreemfs.osdsel_policy_id"));

            if (attrs.containsKey("xtreemfs.free_space"))
                System.out.format(format, "free usable disk space", OutputUtils.formatBytes(Long
                        .valueOf(attrs.get("xtreemfs.free_space"))));

            // if the file does not refer to a directory, render the
            // X-Locations list
            if (ftype != DIR) {

                // because of escape characters, the X-Locations list needs to
                // be parsed in two steps: first, parse the string,
                // then, parse a list from the parsed string
                String s = (String) JSONParser.parseJSON(new JSONString("\""
                    + attrs.get("xtreemfs.locations") + "\""));
                List<Object> l = (List) JSONParser.parseJSON(new JSONString(s));

                System.out.println("\nXtreemFS replica list");
                if (l == null) {
                    System.out.println("   This file does not have any replicas yet.");
                } else {
                    System.out.format(format, "   list version ", l.get(1));
                    List<List> replicas = (List<List>) l.get(0);
                    for (int i = 0; i < replicas.size(); i++) {
                        final Map<String, Object> policy = (Map) replicas.get(i).get(0);
                        final String pStr = policy.get("policy") + "," + policy.get("stripe-size")
                            + "kb," + policy.get("width");
                        System.out.format(format, "   replica " + (i + 1) + " policy", pStr);
                        System.out.format(format, "   replica " + (i + 1) + " OSDs", replicas
                                .get(i).get(1));
                    }
                }

                System.out.println();
            }

        } catch (IOException ex) {
            ex.printStackTrace();
        } catch (InterruptedException ex) {
            ex.printStackTrace();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private static String translatePermissions(Object perms) {
        int perm = ((Long) perms).intValue();

        String pStr = "";

        for (int i = 0; i < 3; i++) {
            if ((perm & 1) != 0) {
                pStr = "x" + pStr;
            } else {
                pStr = "-" + pStr;
            }
            if ((perm & 2) != 0) {
                pStr = "w" + pStr;
            } else {
                pStr = "-" + pStr;
            }
            if ((perm & 4) != 0) {
                pStr = "r" + pStr;
            } else {
                pStr = "-" + pStr;
            }
            perm = perm >> 3;
        }
        return pStr;
    }

}

File: client/tools/xtfs_stat.c
/*  Copyright (c) 2008 Konrad-Zuse-Zentrum fuer Informationstechnik Berlin.

 This file is part of XtreemFS. XtreemFS is part of XtreemOS, a Linux-based
 Grid Operating System, see <http://www.xtreemos.eu> for more details.
 The XtreemOS project has been developed with the financial support of the
 European Commission's IST program under contract #FP6-033576.

 XtreemFS is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 2 of the License, or (at your option)
 any later version.

 XtreemFS is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with XtreemFS. If not, see <http://www.gnu.org/licenses/>.
 */
/*
 * AUTHORS: Björn Kolbeck (ZIB)
 */
#include <stdio.h>
#include <stdlib.h>

#include <sys/types.h>
#ifndef _WIN32
#include <sys/xattr.h>
#endif

#include <json.h>
#include <json_object.h>

size_t getxattr_wrapper(char *path, char *name, char* value, size_t length) {
#ifdef __APPLE__
	return getxattr(path,name,value,length,0,0);
#else
	return getxattr(path,name,value,length);
#endif
}

/*
 *
 */
int main(int argc, char** argv) {

    char *path = argv[1];
    char value[10*1024];
    struct json_object *xloc, *rlist, *replica, *policy, *osds;

    if (argc != 2) {
        printf("usage: %s <filename>\n\n",argv[0]);
        return 1;
    }

    //fetch xtreemfs.url
    ssize_t attrlen = getxattr_wrapper(path,"xtreemfs.url",&value,sizeof(value));
    if (attrlen < 0) {
        printf("ERROR: Cannot retrieve XtreemFS specific file information.\n");
        printf("       The file is probably not part of an XtreemFS volume.\n\n");
        return 1;
    }

    printf("%-18s %s\n","filename",path);

    value[attrlen] = 0;
    printf("%-18s %s\n","XtreemFS URI",(char*)&value);


    //fetch xtreemfs.fileID
    attrlen = getxattr_wrapper(path,"xtreemfs.file_id",&value,sizeof(value));
    value[attrlen] = 0;
    printf("%-18s %s\n","XtreemFS fileID",(char*)&value);

    //fetch xtreemfs.objectType
    attrlen = getxattr_wrapper(path,"xtreemfs.object_type",&value,sizeof(value));
    char *typeName = "";
    char objectType = value[0];
    if (value[0] == '1')
        typeName = "regular file";
    else if (value[0] == '2')
        typeName = "directory";
    else
        typeName = "symlink";

    printf("%-18s %s\n","object type",typeName);

    //fetch xtreemfs.fileID
    attrlen = getxattr_wrapper(path,"xtreemfs.owner",&value,sizeof(value));
    value[attrlen] = 0;
    printf("%-18s %s\n","owner",(char*)&value);

    //fetch xtreemfs.fileID
    attrlen = getxattr_wrapper(path,"xtreemfs.group",&value,sizeof(value));
    value[attrlen] = 0;
    printf("%-18s %s\n","group",(char*)&value);

    //fetch xtreemfs.fileID
    if (objectType == '1') {
        attrlen = getxattr_wrapper(path,"xtreemfs.read_only",&value,sizeof(value));
        if (attrlen > 0) {
            value[attrlen] = 0;
            printf("%-18s %s\n","read-only",(char*)&value);
        } else {
            printf("%-18s %s\n","read-only","no");
        }
    }

    if (objectType == '2') {
        attrlen = getxattr_wrapper(path,"xtreemfs.default_sp",&value,sizeof(value));
        if (attrlen > 0) {
            value[attrlen] = 0;
            printf("%-18s %s\n","default SP",(char*)&value);
        } else {
            printf("%-18s %s\n","default SP","none");
        }
    }


    attrlen = getxattr_wrapper(path,"xtreemfs.ac_policy_id",&value,sizeof(value));
    if (attrlen > 0) {
        value[attrlen] = 0;
        printf("%-25s %s\n","access control policy ID",(char*)&value);
    }

    attrlen = getxattr_wrapper(path,"xtreemfs.osdsel_policy_id",&value,sizeof(value));
    if (attrlen > 0) {
        value[attrlen] = 0;
        printf("%-25s %s\n","OSD selection policy ID",(char*)&value);
    }

    attrlen = getxattr_wrapper(path,"xtreemfs.osdsel_policy_args",&value,sizeof(value));
	if (attrlen > 0) {
		value[attrlen] = 0;
		printf("%-25s %s\n","OSD selection policy arguments",(char*)&value);
	}

    attrlen = getxattr_wrapper(path,"xtreemfs.free_space",&value,sizeof(value));
    if (attrlen > 0) {
        value[attrlen] = 0;
        long long free = strtoll((char*)&value,NULL,10);
        if (free > 1024*1024*1024) {
            free = free/(1024*1024*1024);
            printf("%-25s %lld GB\n","free usable disk space",free);
        } else if (free > 1024*1024) {
            free = free/(1024*1024);
            printf("%-25s %lld MB\n","free usable disk space",free);
        } else if (free > 1024) {
            free = free/1024;
            printf("%-25s %lld kB\n","free usable disk space",free);
        } else {
            printf("%-25s %lld bytes\n","free usable disk space",free);
        }
    }

    //print Locations list
    attrlen = getxattr_wrapper(path,"xtreemfs.locations",&value,sizeof(value));
    if (attrlen > 0) {
        value[attrlen] = 0;

        //printf("attrvalue = %s",&value);

        xloc = json_tokener_parse(value);
        if (is_error(xloc)) {
            printf("ERROR: cannot parse server response.\n\n");
            return 1;
        }

        printf("%-18s %lld\n","replica list ver.",json_object_get_int(json_object_array_get_idx(xloc,1)));
        rlist = json_object_array_get_idx(xloc,0);

        int numReplicas = json_object_array_length(rlist);
        int i,repl,numOsds;

        for (i = 0; i < numReplicas; i++) {
            replica = json_object_array_get_idx(rlist,i);
            policy = json_object_array_get_idx(replica,0);
            char *policyName = json_object_get_string(json_object_object_get(policy,"policy"));
            int   stripeSize = json_object_get_int(json_object_object_get(policy,"stripe-size"));
            int   width = json_object_get_int(json_object_object_get(policy,"width"));
            osds = json_object_array_get_idx(replica,1);
            numOsds = json_object_array_length(osds);
            printf("%-18s %d","replica",i+1);
            printf(" ( policy: %s, width: %d, stripe-size: %dkB )\n",policyName,width,stripeSize);
            for (repl = 0; repl < numOsds; repl++) {
                printf("      OSD %-3d      %s\n",(repl+1),json_object_get_string(json_object_array_get_idx(osds,repl)));
            }
        }
    }
    return (EXIT_SUCCESS);
}

